File: packages/react-native-uploadthing-background/ios/BackgroundUploadStore.swift
import Foundation

private enum StoredBackgroundUploadTaskStatus: String, Codable {
  case queued
  case uploading
  case completed
  case failed
  case cancelled

  var publicStatus: BackgroundUploadTaskStatus {
    switch self {
    case .queued:
      return .queued
    case .uploading:
      return .uploading
    case .completed:
      return .completed
    case .failed:
      return .failed
    case .cancelled:
      return .cancelled
    }
  }
}

private struct StoredBackgroundUploadHeader: Codable {
  let key: String
  let value: String
}

struct StoredBackgroundUploadTaskRecord: Codable {
  let taskId: String
  var status: StoredBackgroundUploadTaskStatus
  let url: String
  let fileUri: String
  let fileName: String
  let mimeType: String
  var bytesSent: Double
  var totalBytes: Double
  var responseCode: Double?
  var responseBody: String?
  var errorMessage: String?
  let createdAt: Double
  var updatedAt: Double
  let method: String
  let headers: [StoredBackgroundUploadHeader]
  var sessionTaskIdentifier: Int?
  var multipartFilePath: String?
  let notificationTitle: String?
  let notificationBody: String?

  init(
    request: BackgroundUploadRequest,
    totalBytes: Double,
    sessionTaskIdentifier: Int?,
    multipartFilePath: String?
  ) {
    let now = Date().timeIntervalSince1970 * 1000

    taskId = request.taskId
    status = .queued
    url = request.url
    fileUri = request.fileUri
    fileName = request.fileName
    mimeType = request.mimeType
    bytesSent = 0
    self.totalBytes = totalBytes
    responseCode = nil
    responseBody = nil
    errorMessage = nil
    createdAt = now
    updatedAt = now
    method = request.method ?? "PUT"
    headers = (request.headers ?? []).map {
      StoredBackgroundUploadHeader(key: $0.key, value: $0.value)
    }
    self.sessionTaskIdentifier = sessionTaskIdentifier
    self.multipartFilePath = multipartFilePath
    notificationTitle = request.notificationTitle
    notificationBody = request.notificationBody
  }

  func asTask() -> BackgroundUploadTask {
    BackgroundUploadTask(
      taskId: taskId,
      status: status.publicStatus,
      url: url,
      fileUri: fileUri,
      fileName: fileName,
      mimeType: mimeType,
      bytesSent: bytesSent,
      totalBytes: totalBytes,
      responseCode: responseCode,
      responseBody: responseBody,
      errorMessage: errorMessage,
      createdAt: createdAt,
      updatedAt: updatedAt
    )
  }
}

final class BackgroundUploadStore {
  static let shared = BackgroundUploadStore()

  private let queue = DispatchQueue(label: "uploadthing.background.store")
  private let encoder = JSONEncoder()
  private let decoder = JSONDecoder()
  private var cachedRecords: [String: StoredBackgroundUploadTaskRecord]?

  private init() {}

  func activeTaskCount() -> Int {
    queue.sync {
      recordsLocked().values.filter {
        $0.status == .queued || $0.status == .uploading
      }.count
    }
  }

  func listTasks() -> [BackgroundUploadTask] {
    queue.sync {
      recordsLocked()
        .values
        .sorted(by: { $0.createdAt < $1.createdAt })
        .map { $0.asTask() }
    }
  }

  func task(taskId: String) -> BackgroundUploadTask? {
    queue.sync {
      recordsLocked()[taskId]?.asTask()
    }
  }

  func record(taskId: String) -> StoredBackgroundUploadTaskRecord? {
    queue.sync {
      recordsLocked()[taskId]
    }
  }

  func record(forSessionTaskIdentifier taskIdentifier: Int) -> StoredBackgroundUploadTaskRecord? {
    queue.sync {
      recordsLocked().values.first(where: { $0.sessionTaskIdentifier == taskIdentifier })
    }
  }

  func upsert(_ record: StoredBackgroundUploadTaskRecord) {
    queue.sync {
      var records = recordsLocked()
      records[record.taskId] = record
      persistLocked(records)
    }
  }

  @discardableResult
  func update(
    taskId: String,
    mutate: (inout StoredBackgroundUploadTaskRecord) -> Void
  ) -> StoredBackgroundUploadTaskRecord? {
    queue.sync {
      var records = recordsLocked()
      guard var record = records[taskId] else {
        return nil
      }

      mutate(&record)
      record.updatedAt = Date().timeIntervalSince1970 * 1000
      records[taskId] = record
      persistLocked(records)
      return record
    }
  }

  @discardableResult
  func remove(taskId: String) -> StoredBackgroundUploadTaskRecord? {
    queue.sync {
      var records = recordsLocked()
      let removed = records.removeValue(forKey: taskId)
      persistLocked(records)
      return removed
    }
  }

  private func recordsLocked() -> [String: StoredBackgroundUploadTaskRecord] {
    if let cachedRecords {
      return cachedRecords
    }

    do {
      let data = try Data(contentsOf: storageURL())
      let decoded = try decoder.decode(
        [String: StoredBackgroundUploadTaskRecord].self,
        from: data
      )
      cachedRecords = decoded
      return decoded
    } catch {
      cachedRecords = [:]
      return [:]
    }
  }

  private func persistLocked(_ records: [String: StoredBackgroundUploadTaskRecord]) {
    cachedRecords = records

    do {
      let data = try encoder.encode(records)
      let url = try storageURL()
      try FileManager.default.createDirectory(
        at: url.deletingLastPathComponent(),
        withIntermediateDirectories: true,
        attributes: nil
      )
      try data.write(to: url, options: .atomic)
    } catch {
      NSLog(
        "[UploadthingBackground] Failed to persist background upload store: \(error.localizedDescription)"
      )
    }
  }

  private func storageURL() throws -> URL {
    let directory = try FileManager.default.url(
      for: .applicationSupportDirectory,
      in: .userDomainMask,
      appropriateFor: nil,
      create: true
    )

    return directory
      .appendingPathComponent("uploadthing-background", isDirectory: true)
      .appendingPathComponent("tasks.json", isDirectory: false)
  }
}

File: packages/react-native-uploadthing-background/android/src/main/java/com/margelo/nitro/uploadthingbackground/BackgroundUploadStore.kt
package com.margelo.nitro.uploadthingbackground

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

internal object BackgroundUploadStore {
  private const val PREFS_NAME = "uploadthing_background_uploads"
  private const val TASKS_KEY = "tasks"
  private val lock = Any()

  fun activeTaskCount(context: Context): Int = synchronized(lock) {
    readAll(context).values.count {
      it.status == BackgroundUploadTaskStatus.QUEUED ||
        it.status == BackgroundUploadTaskStatus.UPLOADING
    }
  }

  fun listTasks(context: Context): Array<BackgroundUploadTask> = synchronized(lock) {
    readAll(context)
      .values
      .sortedBy { it.createdAt }
      .map { it.toPublicTask() }
      .toTypedArray()
  }

  fun getTask(context: Context, taskId: String): BackgroundUploadTask? = synchronized(lock) {
    readAll(context)[taskId]?.toPublicTask()
  }

  fun getRecord(context: Context, taskId: String): StoredBackgroundUploadTaskRecord? =
    synchronized(lock) {
      readAll(context)[taskId]
    }

  fun upsert(context: Context, record: StoredBackgroundUploadTaskRecord) = synchronized(lock) {
    val records = readAll(context).toMutableMap()
    records[record.taskId] = record
    persist(context, records)
  }

  fun update(
    context: Context,
    taskId: String,
    mutate: (StoredBackgroundUploadTaskRecord) -> StoredBackgroundUploadTaskRecord,
  ): StoredBackgroundUploadTaskRecord? = synchronized(lock) {
    val records = readAll(context).toMutableMap()
    val existing = records[taskId] ?: return null
    val updated = mutate(existing).copy(updatedAt = nowMs())
    records[taskId] = updated
    persist(context, records)
    updated
  }

  fun claimForAttempt(
    context: Context,
    taskId: String,
    attemptId: String,
    totalBytes: Double,
  ): StoredBackgroundUploadTaskRecord? = synchronized(lock) {
    val records = readAll(context).toMutableMap()
    val existing = records[taskId] ?: return null
    if (
      existing.status == BackgroundUploadTaskStatus.COMPLETED ||
      existing.status == BackgroundUploadTaskStatus.FAILED ||
      existing.status == BackgroundUploadTaskStatus.CANCELLED
    ) {
      return null
    }

    val updated = existing.copy(
      status = BackgroundUploadTaskStatus.UPLOADING,
      totalBytes = totalBytes,
      bytesSent = 0.0,
      errorMessage = null,
      attemptId = attemptId,
      updatedAt = nowMs(),
    )
    records[taskId] = updated
    persist(context, records)
    updated
  }

  fun updateForAttempt(
    context: Context,
    taskId: String,
    attemptId: String,
    mutate: (StoredBackgroundUploadTaskRecord) -> StoredBackgroundUploadTaskRecord,
  ): StoredBackgroundUploadTaskRecord? = synchronized(lock) {
    val records = readAll(context).toMutableMap()
    val existing = records[taskId] ?: return null
    if (existing.attemptId != attemptId) {
      return null
    }

    val updated = mutate(existing).copy(updatedAt = nowMs())
    records[taskId] = updated
    persist(context, records)
    updated
  }

  fun remove(context: Context, taskId: String): StoredBackgroundUploadTaskRecord? =
    synchronized(lock) {
      val records = readAll(context).toMutableMap()
      val removed = records.remove(taskId)
      persist(context, records)
      removed
    }

  private fun readAll(context: Context): Map<String, StoredBackgroundUploadTaskRecord> {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    val raw = prefs.getString(TASKS_KEY, null) ?: return emptyMap()
    val json = JSONObject(raw)
    val records = mutableMapOf<String, StoredBackgroundUploadTaskRecord>()

    json.keys().forEach { key ->
      records[key] = StoredBackgroundUploadTaskRecord.fromJson(json.getJSONObject(key))
    }
    return records
  }

  private fun persist(
    context: Context,
    records: Map<String, StoredBackgroundUploadTaskRecord>,
  ) {
    val json = JSONObject()
    for ((taskId, record) in records) {
      json.put(taskId, record.toJson())
    }

    context
      .getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
      .edit()
      .putString(TASKS_KEY, json.toString())
      .apply()
  }

  private fun nowMs(): Double = System.currentTimeMillis().toDouble()
}

internal data class StoredBackgroundUploadTaskRecord(
  val taskId: String,
  val status: BackgroundUploadTaskStatus,
  val url: String,
  val fileUri: String,
  val fileName: String,
  val mimeType: String,
  val bytesSent: Double,
  val totalBytes: Double,
  val responseCode: Double?,
  val responseBody: String?,
  val errorMessage: String?,
  val createdAt: Double,
  val updatedAt: Double,
  val attemptId: String?,
  val method: String,
  val headers: List<BackgroundUploadHeader>,
  val notificationTitle: String?,
  val notificationBody: String?,
) {
  fun toPublicTask(): BackgroundUploadTask =
    BackgroundUploadTask(
      taskId = taskId,
      status = status,
      url = url,
      fileUri = fileUri,
      fileName = fileName,
      mimeType = mimeType,
      bytesSent = bytesSent,
      totalBytes = totalBytes,
      responseCode = responseCode,
      responseBody = responseBody,
      errorMessage = errorMessage,
      createdAt = createdAt,
      updatedAt = updatedAt,
    )

  fun toJson(): JSONObject {
    val json = JSONObject()
    json.put("taskId", taskId)
    json.put("status", status.name)
    json.put("url", url)
    json.put("fileUri", fileUri)
    json.put("fileName", fileName)
    json.put("mimeType", mimeType)
    json.put("bytesSent", bytesSent)
    json.put("totalBytes", totalBytes)
    json.put("responseCode", responseCode)
    json.put("responseBody", responseBody)
    json.put("errorMessage", errorMessage)
    json.put("createdAt", createdAt)
    json.put("updatedAt", updatedAt)
    json.put("attemptId", attemptId)
    json.put("method", method)
    json.put(
      "headers",
      JSONArray().apply {
        headers.forEach { header ->
          put(
            JSONObject()
              .put("key", header.key)
              .put("value", header.value),
          )
        }
      },
    )
    json.put("notificationTitle", notificationTitle)
    json.put("notificationBody", notificationBody)
    return json
  }

  companion object {
    fun fromRequest(request: BackgroundUploadRequest): StoredBackgroundUploadTaskRecord {
      val now = System.currentTimeMillis().toDouble()

      return StoredBackgroundUploadTaskRecord(
        taskId = request.taskId,
        status = BackgroundUploadTaskStatus.QUEUED,
        url = request.url,
        fileUri = request.fileUri,
        fileName = request.fileName,
        mimeType = request.mimeType,
        bytesSent = 0.0,
        totalBytes = 0.0,
        responseCode = null,
        responseBody = null,
        errorMessage = null,
        createdAt = now,
        updatedAt = now,
        attemptId = null,
        method = request.method ?: "PUT",
        headers = request.headers?.toList() ?: emptyList(),
        notificationTitle = request.notificationTitle,
        notificationBody = request.notificationBody,
      )
    }

    fun fromJson(json: JSONObject): StoredBackgroundUploadTaskRecord {
      val now = System.currentTimeMillis().toDouble()
      val headersJson = json.optJSONArray("headers") ?: JSONArray()
      val headers = buildList {
        for (index in 0 until headersJson.length()) {
          val item = headersJson.getJSONObject(index)
          add(
            BackgroundUploadHeader(
              key = item.getString("key"),
              value = item.getString("value"),
            ),
          )
        }
      }

      return StoredBackgroundUploadTaskRecord(
        taskId = json.getString("taskId"),
        status = BackgroundUploadTaskStatus.valueOf(json.getString("status")),
        url = json.getString("url"),
        fileUri = json.getString("fileUri"),
        fileName = json.getString("fileName"),
        mimeType = json.getString("mimeType"),
        bytesSent = json.optDouble("bytesSent", 0.0),
        totalBytes = json.optDouble("totalBytes", 0.0),
        responseCode = json.optDouble("responseCode").takeUnless { json.isNull("responseCode") },
        responseBody = json.optString("responseBody").takeUnless { json.isNull("responseBody") },
        errorMessage = json.optString("errorMessage").takeUnless { json.isNull("errorMessage") },
        createdAt = json.optDouble("createdAt", now),
        updatedAt = json.optDouble("updatedAt", now),
        attemptId = json.optString("attemptId").takeUnless { json.isNull("attemptId") },
        method = json.optString("method", "PUT"),
        headers = headers,
        notificationTitle = json.optString("notificationTitle")
          .takeUnless { json.isNull("notificationTitle") },
        notificationBody = json.optString("notificationBody")
          .takeUnless { json.isNull("notificationBody") },
      )
    }
  }
}
